boot/freeldr/freeldr/arch/uefi/uefidisk.c

```diff
@@ -0,0 +1,591 @@
+/*
+ * PROJECT:     FreeLoader UEFI Support
+ * LICENSE:     GPL-2.0-or-later (https://spdx.org/licenses/GPL-2.0-or-later)
+ * PURPOSE:     Disk Access Functions
+ * COPYRIGHT:   Copyright 2022 Justin Miller <justinmiller100@gmail.com>
+ */
+
+/* INCLUDES ******************************************************************/
+
+#include <uefildr.h>
+
+#include <debug.h>
+DBG_DEFAULT_CHANNEL(WARNING);
+
+#define TAG_HW_RESOURCE_LIST    'lRwH'
+#define TAG_HW_DISK_CONTEXT     'cDwH'
+#define FIRST_BIOS_DISK 0x80
+#define FIRST_PARTITION 1
+
+typedef struct tagDISKCONTEXT
+{
+    UCHAR DriveNumber;
+    ULONG SectorSize;
+    ULONGLONG SectorOffset;
+    ULONGLONG SectorCount;
+    ULONGLONG SectorNumber;
+} DISKCONTEXT;
+
+typedef struct _INTERNAL_UEFI_DISK
+{
+    UCHAR ArcDriveNumber;
+    UCHAR NumOfPartitions;
+    UCHAR UefiRootNumber;
+    BOOLEAN IsThisTheBootDrive;
+} INTERNAL_UEFI_DISK, *PINTERNAL_UEFI_DISK;
+
+/* GLOBALS *******************************************************************/
+
+extern EFI_SYSTEM_TABLE* GlobalSystemTable;
+extern EFI_HANDLE GlobalImageHandle;
+extern EFI_HANDLE PublicBootHandle; /* Freeldr itself */
+
+/* Made to match BIOS */
+UCHAR FrldrBootDrive;
+ULONG FrldrBootPartition;
+SIZE_T DiskReadBufferSize;
+PVOID DiskReadBuffer;
+UCHAR PcBiosDiskCount;
+PVOID Buffer;
+
+static const CHAR Hex[] = "0123456789abcdef";
+static CHAR PcDiskIdentifier[32][20];
+
+/* Uefi Specifc */
+UINT64* LoaderFileSize;
+ULONG UefiBootRootIdentifier;
+ULONG OffsetToBoot;
+ULONG PublicBootArcDisk;
+INTERNAL_UEFI_DISK* InternalUefiDisk = NULL;
+EFI_GUID bioGuid = BLOCK_IO_PROTOCOL;
+EFI_BLOCK_IO* bio;
+EFI_HANDLE* handles = NULL;
+
+/* FUNCTIONS *****************************************************************/
+
+PCHAR
+GetHarddiskIdentifier(UCHAR DriveNumber)
+{
+    TRACE("GetHarddiskIdentifier: DriveNumber: %d\n", DriveNumber);
+    return PcDiskIdentifier[DriveNumber - FIRST_BIOS_DISK];
+}
+
+static LONG lReportError = 0; // >= 0: display errors; < 0: hide errors.
+
+LONG
+DiskReportError(BOOLEAN bShowError)
+{
+    /* Set the reference count */
+    if (bShowError) ++lReportError;
+    else            --lReportError;
+    return lReportError;
+}
+
+static
+BOOLEAN
+UefiGetBootPartitionEntry(
+    IN UCHAR DriveNumber,
+    OUT PPARTITION_TABLE_ENTRY PartitionTableEntry,
+    OUT PULONG BootPartition)
+{
+    ULONG PartitionNum;
+    TRACE("UefiGetBootPartitionEntry: DriveNumber: %d\n", DriveNumber - FIRST_BIOS_DISK);
+    /* UefiBootRoot is the offset into the array of handles where the raw disk of the boot drive is */
+    PartitionNum = (OffsetToBoot - UefiBootRootIdentifier);
+    /* Partitions start with 1 in ARC, but, UEFI root drive identifier is also the first partition */
+    if (PartitionNum == 0)
+    {
+        ERR("Boot PartitionNumber is 0\n");
```

Comment:
What else should be done in this case?

Reply:
I wasn't sure, this shouldn't ever happen and after i got the disk code actually working it hasn't happened. I can remove it? or maybe add an assertion

Reply:
If you get PartitionNum == 0 this means OffsetToBoot == UefiBootRootIdentifier. If this can happen then you'll get zero. Otherwise if it's never to be expected (check how this is retrieved, I guess?), then an assert should be OK to have.